libs/shmpipe/makefile

```diff
@@ -0,0 +1,9 @@
+libshmpipe.so: shmpipe.c
+	gcc -pthread -o libshmpipe.so -shared -g -O2 shmpipe.c
+
+shmpipe: shmpipe.c
+	gcc -pthread -o shmpipe -DSHMEM_PIPE_TEST -g -O2 shmpipe.c -lrt
+
+install:
+	sudo cp libshmpipe.so /usr/lib
+	sudo cp shmpipe.h /usr/include
```

Comment:
Using sudo here is not optimal for our development.

I wonder if this could be compiled with `cc` crate in the `libs/shmpipe` and that static library be reused...? Well, that would create a circular dependency between:
- rust build needs the headers from postgres build
- pgxn needs the static library from rust build

Which would be the same with any other more rustified shmpipe implementation..

Alternative is to just build it separatedly as part of each pgxn build. I wonder if that presents some hazards.

Out of interest, is it needed as a shared library or could it be a static as well?

Reply:
As I wrote in the comments, I am absolutely sure that building and installation procedure for `shmpipe` needs to be changed. I am not sure whether it should be built by cargo or by makefile. What we need:
1. Build shmem shared or static library
2. Add this library to the linker when building postgres
3. Refer to this library in Rust wrapper for shmpipe.

I will be glad to here suggestions what is the most optimal way of doing it.

Reply:
After writing the above comment it occured to me that of course it could be built as a separate crate and used by all without issues, so there's likely no problems there.

Reply:
I first of all want to receive some feedback on the proposed approach with shmpipe. There are some questions:
1. Should it be implemented in C with unsafe calls fro Rust (as it is done now), or in C (assuming that it is possible to call Rust library from C) or in C/rust (if both are using the same system calls for synchronization or not using them at all: spinlocks/atomics/...)?
2.  Should this pipe implement multiplexing (multiple producers single consumer) or should it be classical end-to-end channel with multiplexing implemented at Rust level? 
3. `shmem_pipe_process_request` (the only function called by Rust code now) may block execution of current thread for some time. Can it cause some problems because this functions is called from Tokio task?

And some known problems:
1. How to handle walredo errors. Previously them are extracted fro walredo stderr. In principle we can do the same here. Or return error code as part of response. Actually I am not se that it is needed, because walredo errors should be considered as critical and non-recoverable.
2. How to properly shutdown walredo process (it is also valid for the current implementation)
3. Can/should we eliminate lazy loading of walredo process? (See https://github.com/neondatabase/neon/issues/2400)
4. Is it possible to avoid extra copying of data. Right now pageserver materilizes deltas and base image in Rust memory, copy them to shared memory, then walredo process copies data to its  private memory, reconstruct page in its private memory and copies it to shared memory, from where it is copied by pageserver to Rust memory. So ~8k of data is copied four times per wal-redo request.

Reply:
Discussed with @problame, I had a shared memory misunderstanding when I was thinking of this:

> 3. Theoretically the shmem_pipe_process_request could be made async as well, so that the C-side would "wake up" the rust future waiting, if any. Given the other blocking issues and some open tokio issues this would probably be slower however.

And I haven't figured out how to otherwise do it (without signals or stdio).

Reply:
In Postgres latches are implemented using ... pipes. When you need to wait - you just do poll on pipe descriptor.
And to signal it you write dummy byte into the pipe.

When I first saw this implementation, I decided that it should be extremely slow (comparing with system synchronization primitives like semaphores or pthread CV). But my experiments show that performance is comparable in all cases (looks like it is determined by number of syscalls, doesn't matter which syscall is it).

So it is really possible to create async shmem pipe. The challenge is how to make it usable both from C and Rust.
But still not sure of it is needed.

Reply:
Out of interest, found out that ipc-channel runs a "Router" thread, which is used in some wrapper crate to provide wakeups. But resolving this thread, because async is in our interest.